### PR TITLE
[extensions/observer/cfgardenobserver] Implement cfgardenobserver

### PR DESCRIPTION
**Pull request by tomasmota**
**Description:** 
First Component PR: #33727
This is the second PR for adding the cfgardenobserver, with the first suggested implementation. There are definitely some decisions made that require feedback, such as adding the CloudFoundry application labels to the Endpoint labels, and the decision to use the `Container` EndpointType at all.

**Link to tracking Issue:** #33618 

**Testing:** Unit testing of config and extension

**Documentation:** Updated readme with new configuration and endpoints

### CODE REVIEW

**Comment by crobert-1** [comment]
```
Generated code is out of date, please run "make generate" and commit the changes in this PR.
```

File: extension/observer/cfgardenobserver/README.md
Comment (crobert-1) [inline extension/observer/cfgardenobserver/README.md:25]: Just a nit, since we have the table below that explains each configuration option, we probably don't need a comment for each option in this config example.

File: extension/observer/cfgardenobserver/config.go
Comment (crobert-1) [inline extension/observer/cfgardenobserver/config.go:118]: We should make as many of these private as possible.

File: extension/observer/cfgardenobserver/config.go
Comment (crobert-1) [inline extension/observer/cfgardenobserver/config.go:49]: The error messages don't seem to align with the README table for config options. The table states that these options are simply required, without any conditional placed on `include_app_labels`. I may be missing something, but which one is correct?

File: extension/observer/cfgardenobserver/config.go
Comment (crobert-1) [inline extension/observer/cfgardenobserver/config.go:74]:
```suggestion
		return fmt.Errorf("configuration option `auth_type` must be set to one of the following values: [user_pass, client_credentials, token]. Specified value: %s", c.AuthType)
```
It would be good to show that it's required, what values are valid, and what the user specified here.

File: extension/observer/cfgardenobserver/README.md
Comment (crobert-1) [inline extension/observer/cfgardenobserver/README.md:55]:
```suggestion
| refresh_interval            | string | 1m                                | Determines how often to look for changes in endpoints.            |
```
Nit

File: extension/observer/cfgardenobserver/config_test.go
Comment (crobert-1) [inline extension/observer/cfgardenobserver/config_test.go:120]: Just a thought: I see some value in checking specific error messages here, but I'm also concerned that this will cause a lot of noise when needing to update error messages for clarity etc. We could simply switch to a boolean that shows whether we expect and error or not. Your call, whatever you think makes more sense 👍

File: extension/observer/cfgardenobserver/extension.go
Comment (crobert-1) [inline extension/observer/cfgardenobserver/extension.go:33]: [Contexts shouldn't be stored in `structs`.](https://go.dev/blog/context-and-structs) Is there some other way to handle the necessary functionality?

File: extension/observer/cfgardenobserver/extension.go
Comment (crobert-1) [inline extension/observer/cfgardenobserver/extension.go:334]: I don't really think having this in a separate function provides much value. Callers still have to check the returned error, so it just ends up being a function call vs. accessing the map directly.

File: extension/observer/cfgardenobserver/extension.go
Comment (crobert-1) [inline extension/observer/cfgardenobserver/extension.go:335]: It maybe helpful to have a central variable for `network.app_id`.

File: extension/observer/cfgardenobserver/extension.go
Comment (crobert-1) [inline extension/observer/cfgardenobserver/extension.go:337]: This error message may not be super helpful to users. What's the relevance to the end user that they could understand?

File: extension/observer/cfgardenobserver/extension.go
Comment (crobert-1) [inline extension/observer/cfgardenobserver/extension.go:175]: Maybe use a central variable for "active". I see the cloud foundry golang package doesn't have one, so we can just declare our own.

File: extension/observer/cfgardenobserver/extension.go
Comment (crobert-1) [inline extension/observer/cfgardenobserver/extension.go:190]: Maybe we can use a central variable for "network.ports"

File: extension/observer/cfgardenobserver/extension.go
Comment (crobert-1) [inline extension/observer/cfgardenobserver/extension.go:202]:
```suggestion
			g.logger.Error("error fetching application", zap.Error(err))
```
Nit

File: extension/observer/cfgardenobserver/extension.go
Comment (crobert-1) [inline extension/observer/cfgardenobserver/extension.go:209]: We shouldn't have `port` declared twice here. Maybe the variable declared in the `for` line could be `portString` or `portUnformatted`. Something else may be better.

Comment (crobert-1) [inline extension/observer/cfgardenobserver/extension.go:209]: `err` is already declared, so we don't need the `:=`, maybe we declare `port` above and convert this to a plain `=`

File: extension/observer/cfgardenobserver/extension.go
Comment (crobert-1) [inline extension/observer/cfgardenobserver/extension.go:278]: Would it make sense to make `log_config` a central variable as well?

Follow up question for this PR as a whole: Does the cf golang package define any of these properties as constants somewhere that we could use, instead of defining them all ourselves?

File: extension/observer/cfgardenobserver/extension.go
Comment (crobert-1) [inline extension/observer/cfgardenobserver/extension.go:255]:
```suggestion
// has contents that look like the following JSON encoded string:
```

File: extension/observer/cfgardenobserver/extension.go
Comment (crobert-1) [inline extension/observer/cfgardenobserver/extension.go:289]: Same request, make `tags` a central variable.

**Review by crobert-1 — COMMENTED** [review]
Mostly just nits.

_I haven't gotten a chance to look through tests yet, I wanted to get a review out sooner rather than later. Hope it's helpful_ 👍

File: extension/observer/cfgardenobserver/config_test.go
Comment (tomasmota) [reply extension/observer/cfgardenobserver/config_test.go:120]: Had the same thought, but with the different conditionals it's pretty easy to get an error by getting into the wrong branch, so I think we can keep the messages for now and remove later if it turns out to be noisy.

File: extension/observer/cfgardenobserver/extension.go
Comment (tomasmota) [reply extension/observer/cfgardenobserver/extension.go:337]: replaced with `return fmt.Errorf("container properties do not have a '%s' field, required to fetch App labels", propertiesAppIDKey)
`

File: extension/observer/cfgardenobserver/config.go
Comment (tomasmota) [reply extension/observer/cfgardenobserver/config.go]: I have made the consts private, should AuthType also be private?

File: extension/observer/cfgardenobserver/config.go
Comment (tomasmota) [reply extension/observer/cfgardenobserver/config.go]: You're right, I added the `include_app_labels` option in the end after opening the PR, and forgot to adjust the README

File: extension/observer/cfgardenobserver/extension.go
Comment (tomasmota) [reply extension/observer/cfgardenobserver/extension.go]: I have made all of these central variables now. 

The [garden client](https://pkg.go.dev/code.cloudfoundry.org/garden), which is use to fetch this information, does not provide any constants for these. As far as I can tell these are just properties created by [guardian](https://github.com/cloudfoundry/guardian/) when creating the containers, which doesn't seem to have them as constants either.

File: extension/observer/cfgardenobserver/extension.go
Comment (tomasmota) [reply extension/observer/cfgardenobserver/extension.go:40]: Good point, I honestly just did the same as the dockerobserver. Will try to find an alternative.

**Comment by tomasmota** [comment]
> Mostly just nits.
> 
> _I haven't gotten a chance to look through tests yet, I wanted to get a review out sooner rather than later. Hope it's helpful_ 👍

I appreciate that, have addressed all the issues except for the context thing, which I'll try to change later.

File: extension/observer/cfgardenobserver/config.go
Comment (crobert-1) [reply extension/observer/cfgardenobserver/config.go]: Yes, we should make `AuthType` private as well, if possible.

File: extension/observer/cfgardenobserver/README.md
Comment (crobert-1) [inline extension/observer/cfgardenobserver/README.md:53]: What do you think of moving authentication options under a specific config section for each? Something like:
```
auth_type: user_pass
user_pass:
  username: "a"
  password: "..."
```
I'm wondering if it would help clarify that each of these options are only relevant for a specific authentication type. If a config then contains options irrelevant to the `auth_type` chosen, they'd just be ignored. Do you think that would be a more clear way of handling this?

File: extension/observer/cfgardenobserver/config_test.go
Comment (crobert-1) [inline extension/observer/cfgardenobserver/config_test.go:193]: `loadConfig` and `loadConf` are a bit unclear on how they differ and their purpose. Can we rename to make their difference more clear? `loadConf` could be something like `loadConfFromFile`, or something else if you have a better idea.

File: extension/observer/cfgardenobserver/README.md
Comment (tomasmota) [reply extension/observer/cfgardenobserver/README.md:53]: hmm it's a good point. I tried implementing your suggestion and after seeing it I personally feel like it's more clunky to have them in separate sections, I would rather make it clearer in the readme.

The following, for example, looks a bit odd to me:
```
auth_type: token
token:
  access_token: "a"
  refresh_token: "..."
```

Other namings I tried for the token stuff also seem odd to me, but maybe this is just a personal preference. Ultimately I value your opinion more so if you still think it's best of course I will do it that way!

Comment (crobert-1) [reply extension/observer/cfgardenobserver/README.md:53]: That's fair, I can see that being a bit clunky. Would it make sense to put all authentication options under the same header?

```
authentication:
  type: token
  access_token: "a"
  refresh_token: "..."
```
It could be `auth` instead of `authentication`, but do you think that would help? I just want to avoid having such a flat layout when all of these options are related/dependent on each other somehow.

File: extension/observer/cfgardenobserver/README.md
Comment (crobert-1) [inline extension/observer/cfgardenobserver/README.md:54]:
```suggestion
| cloud_foundry.username      | string | none                                                                   | Username (auth_type: user_pass)                                    |
```

File: extension/observer/cfgardenobserver/config.go
Comment (crobert-1) [inline extension/observer/cfgardenobserver/config.go:26]:
```suggestion
	// using the CloudFoundry API.
```
Nit: Config examples are in the README, it's not necessary to include here as well.

**Review by crobert-1 — COMMENTED** [review]
Looking pretty good, just another suggestion on configuration layout, and some nits.

File: extension/observer/cfgardenobserver/README.md
Comment (tomasmota) [reply extension/observer/cfgardenobserver/README.md]: good idea, I've changed it now

**Review by crobert-1 — APPROVED** [review]

**Comment by crobert-1** [comment]
Failing unit test is unrelated to this change, I've filed #35007.

**Comment by tomasmota** [comment]
hey @crobert-1 there is this single prometheus compliance test that keeps failing for more than a week now and I have no clue why. I have tried checking if it was something with dependencies but no luck. Do you have any idea what it could be?

**Comment by crobert-1** [comment]
> hey @crobert-1 there is this single prometheus compliance test that keeps failing for more than a week now and I have no clue why. I have tried checking if it was something with dependencies but no luck. Do you have any idea what it could be?

Hey @tomasmota, sorry about that. It's frequency of https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/35119, unrelated to your changes 👍

**Comment by crobert-1** [comment]
_It looks like some imports are the wrong version again, but I'm adding `ready to merge` as it should be good once that's fixed._ 👍

**Comment by andrzej-stencel** [comment]
Sorry for the delay @tomasmota. Can you fix the conflicts again :sweat:

**Comment by andrzej-stencel** [comment]
@m1rp It looks like this still has conflicts, can you take a look?

**Comment by m1rp** [comment]
i've also done the CLA authorization now

**Comment by atoulme** [comment]
easyCLA still fails, please review

**Comment by MovieStoreGuy** [comment]
Would it be possible to squash some of these commits? Hopefully it should make any branch updates easier :)

**Comment by m1rp** [comment]
EasyCLA has worked this time thankfully!

**Pull request merged**
